system/jlib/jptree.cpp

```diff
@@ -9481,3 +9542,341 @@ void copyPropIfMissing(IPropertyTree & target, const char * targetName, IPropert
+{
+    if (!tree)
+        return;
+    PTree *pt = static_cast<PTree*>(tree);
+    pt->markNameEncoded();
+}
+
+bool isPTreeNameEncoded(const IPropertyTree *tree)
+{
+    if (!tree)
+        return false;
+    const PTree *pt = static_cast<const PTree*>(tree);
+    return pt->isNameEncoded();
+}
+
+void setPTreeAttribute(IPropertyTree *tree, const char *name, const char *value, bool markEncoded)
+{
+    if (!tree || isEmptyString(name))
+        return;
+    PTree *pt = static_cast<PTree*>(tree);
+    pt->setAttribute(name, value, markEncoded);
+}
+
+bool isPTreeAttributeNameEncoded(const IPropertyTree *tree, const char *name)
+{
+    if (!tree || isEmptyString(name))
+        return false;
+    const PTree *pt = static_cast<const PTree*>(tree);
+    return pt->isAttributeNameEncoded(name);
+}
+
+static void decodePTreeNameHexEncoding(StringBuffer &s, const char *&input, unsigned &remaining)
+{
+    byte ch = 0;
+    while (remaining >= 2 && *input!='_')
+    {
+        byte high = hex2num(*input++);
+        ch = (high << 4) | hex2num(*input++);
+        remaining -= 2;
+        s.append(ch);
+    }
+    if (*input=='_')
+    {
+        input++;
+        remaining--;
+    }
+}
+
+static void decodePTreeNameEncoding(StringBuffer &s, const char *&input, unsigned &remaining)
+{
+    if (*input!='_')
+        return;
+    input++;
+    remaining--;
+    if (remaining==0)
+        return;
+    if (*input=='x')
+    {
+        input++;
+        remaining--;
+        decodePTreeNameHexEncoding(s, input, remaining);
+        return;
+    }
+    switch (*input)
+    {
+        case '0': //we only currently use this for empy names (including "@" with no tail)
+            break;
+        case '_':
+            s.append('_');
+            break;
+        case 'a':
+            s.append('@');
+            break;
+        case 'b':
+            s.append('\\');
+            break;
+        case 'f':
+            s.append('/');
+            break;
+        case 'n':
+            s.append('\n');
+            break;
+        case 'Q':
+            s.append('\"');
+            break;
+        case 'r':
+            s.append('\r');
+            break;
+        case 's':
+            s.append(' ');
+            break;
+        case 'q':
+            s.append('\'');
+            break;
+        case 't':
+            s.append('\t');
+            break;
+        default: //shouldn't get here since encoding can't be done externally, but treat as if not encoded
+            s.append('_').append(*input);
+            break;
+    }
+    input++;
+    remaining--;
+}
+
+StringBuffer &decodePtreeName(StringBuffer &s, const char *input, unsigned remaining)
+{
+    if (!input || !remaining)
+        return s;
+    while (remaining)
+    {
+        if ('_'!=*input)
+        {
+            s.append(*input);
+            input++;
+            remaining--;
+        }
+        else
+            decodePTreeNameEncoding(s, input, remaining);
```

Comment:
probably similar elsewhere, but wonder on cost of overhead of repeated calls to StringBuffer::append here.
Could you pre-reserve the estiated output size in the StringBuffer?

Reply:
done, see added ensureCapacity() above.